### ISSUE TITLE
Spike: Try out `createPreloadingMiddleware` to replace settings hydration hook

### ISSUE THREAD

**Issue by louwie17** [woocommerce/woocommerce#32208]
### Description

We are [currently hydrating a bunch](https://github.com/woocommerce/woocommerce-admin/blob/main/client/index.js#L30-L33) of our mutable settings into our @wc/admin/settings datastore by making use of `wcSettings`. This allows a user to retrieve mutable settings two way's, using our data store or making use of `wc.wcSettings.getSetting`. This could add confusion, and the latter is only for static settings.

We should try out the `createPreloadingMiddleware` for any settings that are only used through our settings store (or any settings that are mutable). This is a spike, so feel free to try out other approaches, like a custom REST route or so, and see if this approach would be best (at-least for now).

Here is an example of someone making use of `createPreloadingMiddleware` -> https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5022
And here is the [source code for the middleware](https://github.com/WordPress/gutenberg/blob/trunk/packages/api-fetch/src/middlewares/preloading.js).

### Acceptance Criteria
- Have a sample PR ready making use of `createPreloadingMiddleware`
- Post any feedback in this issue, on why we should (or should not) use this
     - If not, what is a good alternative?

**Comment by psealock** [comment]
@louwie17, this sounds like an interesting idea. Here are just a few thoughts:

We currently try to discourage accessing mutable settings via `wc.wcSettings.getSettings` by throwing an error if accessed directly.

https://github.com/woocommerce/woocommerce-admin/blob/e65aec4cdc67e0e280d4d7722876670f7b0647c1/packages/wc-admin-settings/src/index.js#L46-L50

This system evolved into this dysfunctional state and an improvement would be welcome. Would 3PD be able to access preloaded data if they weren't using an API request?

**Comment by louwie17** [comment]
> We currently try to discourage accessing mutable settings via wc.wcSettings.getSettings by throwing an error if accessed directly.

Yeah this would support this and push people to use the data store instead. This piece of code is only used internally by us though, so won't be of any use to a 3PD, they would be using https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/settings/shared/utils.ts#L18-L26 which doesn't throw that warning.

> This system evolved into this dysfunctional state and an improvement would be welcome. Would 3PD be able to access preloaded data if they weren't using an API request?

Yeah the `preloadingMiddleware` acts almost identical to our hydratingSettings hook, and 3PD's would still be able to access the data through the data store without triggering an API request.
The `preloadingMiddleware` is just more wp store specific and will prevent users from getting it the `wc.wcSettings.getSetting` route.

I am open to other suggestions, although I think this will help keep some of our settings data more consolidated.

**Comment by mattsherman** [comment]
@louwie17 What's the status of this one? Is it still "In Progress"?

**Comment by louwie17** [comment]
> @louwie17 What's the status of this one? Is it still "In Progress"?

Not anymore now that cooldown is over, I unassigned myself and moved it to the Product backlog (not a 100% sure if that is the right column).

**Comment by timmy5685** [comment]
Hi @louwie17 - thanks for such a detailed report, it's really helpful.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.

The bit that needs a closer look: I did not mark this invalid because the exact hydration hook still exists. I did not mark it ready_eng because the codebase has partially adopted preloading middleware and the original spike framing may be obsolete; this needs an architecture decision before PMs can confidently route it.

**Comment by vladolaru** [comment]
I checked this against current trunk, the existing issue thread, and the current WordPress preload APIs. I would keep this open, but re-scope it as a narrow architecture spike rather than route it as a ready implementation issue.

The original hydration path still exists. The main admin entry hydrates `wc_admin` from `window.wcSettings.admin` and then hydrates `general` from `window.wcSettings.admin?.preloadSettings?.general` via [`withSettingsHydration()`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/client/index.tsx#L33-L47). The HOC marks the `getSettings( group )` resolver as started/finished and writes the supplied settings into the store ([source](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/settings/with-settings-hydration.tsx#L14-L65)).

The resolver path is important here. The normal settings resolver fetches [`NAMESPACE + '/settings/' + group`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/settings/resolvers.ts#L58-L67), and in this package `NAMESPACE` is [`/wc-analytics`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/constants.ts#L1-L3). The matching REST controller is `SettingOptions`, which overrides the inherited settings controller namespace to [`wc-analytics`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/SettingOptions.php#L20-L28), and the parent controller registers `/settings/{group_id}` and returns the raw settings array ([route/response source](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-setting-options-v2-controller.php#L40-L153)). So a representative spike should preload `/wc-analytics/settings/general`, unless the PR also changes the resolver.

The current PHP preload producer is still a custom `wcSettings.admin.preloadSettings` path rather than API preloading. `Settings.php` and `Loader.php` read `woocommerce_admin_preload_settings`, instantiate `WC_REST_Setting_Options_V2_Controller`, call `get_group_settings( $group )`, and reduce the result to a simplified `{ [id]: value }` map under `preloadSettings[group]` ([Settings.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Admin/Settings.php#L165-L178), [Loader.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Admin/Loader.php#L374-L386)). That means the current code bypasses the resolver response shape and then hydrates resolver state manually.

The middleware approach is viable in this codebase. WooCommerce already uses `wp.apiFetch.createPreloadingMiddleware()` in Blocks asset data and the Email Editor ([Blocks](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/Assets/AssetDataRegistry.php#L405-L407), [Email Editor](https://github.com/woocommerce/woocommerce/blob/trunk/packages/php/email-editor/src/Engine/class-assets-manager.php#L239-L251)). WordPress also documents the server-side preload helper in [`rest_preload_api_request()`](https://developer.wordpress.org/reference/functions/rest_preload_api_request/) and uses the same `rest_preload_api_request()` plus `createPreloadingMiddleware()` pattern in [`block_editor_rest_api_preload()`](https://developer.wordpress.org/reference/functions/block_editor_rest_api_preload/).

I would not try to replace all `window.wcSettings.admin` hydration in this issue. The useful next step is a focused `general` settings spike:

- Preload `/wc-analytics/settings/general` with `rest_preload_api_request()`.
- Register that response with `wp.apiFetch.createPreloadingMiddleware()`.
- Remove or bypass `withSettingsHydration( 'general', ... )` and let the existing `getSettings( 'general' )` resolver consume the preloaded response.
- Migrate direct `window.wcSettings.admin.preloadSettings.general` reads, including the main admin entry and the current ExPlat/payments readers ([ExPlat](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/explat/src/assignment.ts#L54-L64), [BACS settings](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/client/settings-payments/offline/settings-payments-bacs.tsx#L34-L37), [payment gateways](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx#L82-L85)).
- Compare payload size, PHP runtime/queries, first-render behavior, and resolver state before recommending any broader migration.

Related work has modernized or optimized the surrounding pieces, but did not complete this replacement: #55425 modernized the settings data-store registration, #61818 optimized avoidable settings option queries, and #65505 removed a costly unrelated REST preload from the shared admin settings payload.

So my recommendation is: keep this open as a narrow spike if we want that sample PR and measurement. If we do not want to run the sample, close it explicitly as "current hybrid `wcSettings` + settings-store hydration accepted" rather than leaving it as a broad backlog item.

**Comment by gigitux** [comment]
Given that this was a spike, I’m going to close this issue. If needed, we can reopen it and continue working on it.
